File: src/poison.py
"""Evaluation contamination sentinel — exact identity/feature overlap evidence.

This mechanism detects exact example-ID overlap and exact feature-hash overlap.
A feature hash may represent a similarity bucket produced elsewhere, but this
module does not itself compute semantic or near-duplicate similarity.
"""
from __future__ import annotations

import hashlib
import json
import math
from dataclasses import dataclass
from typing import Sequence


def digest(obj: object) -> str:
    return hashlib.sha256(
        json.dumps(obj, sort_keys=True, separators=(",", ":"), default=str).encode()
    ).hexdigest()


@dataclass(frozen=True)
class Example:
    example_id: str
    feature_hash: str


@dataclass(frozen=True)
class ContaminationReport:
    id_overlap: int
    feature_overlap: int
    feature_overlap_ratio: float
    train_n: int
    eval_n: int
    contaminated: bool
    input_fingerprint: str
    policy_fingerprint: str
    fingerprint: str


class EvalPoisonSentinel:
    """Fail closed when exact overlap exceeds an explicit bounded policy."""

    def __init__(
        self,
        max_id_overlap: int = 0,
        max_feature_overlap_ratio: float = 0.01,
    ):
        if max_id_overlap < 0:
            raise ValueError("max_id_overlap must be non-negative")
        if (
            not math.isfinite(max_feature_overlap_ratio)
            or not 0.0 <= max_feature_overlap_ratio <= 1.0
        ):
            raise ValueError("max_feature_overlap_ratio must be finite and in [0,1]")
        self.max_id_overlap = max_id_overlap
        self.max_feature_overlap_ratio = max_feature_overlap_ratio

    @staticmethod
    def _validate_dataset(name: str, examples: Sequence[Example]) -> None:
        seen_ids: set[str] = set()
        for example in examples:
            if not example.example_id.strip():
                raise ValueError(f"{name} example_id must be non-empty")
            if not example.feature_hash.strip():
                raise ValueError(f"{name} feature_hash must be non-empty")
            if example.example_id in seen_ids:
                raise ValueError(f"duplicate {name} example_id: {example.example_id}")
            seen_ids.add(example.example_id)

    @staticmethod
    def _dataset_fingerprint(examples: Sequence[Example]) -> list[tuple[str, str]]:
        return sorted((example.example_id, example.feature_hash) for example in examples)

    def analyze(
        self,
        train: Sequence[Example],
        eval_set: Sequence[Example],
    ) -> ContaminationReport:
        self._validate_dataset("train", train)
        self._validate_dataset("eval", eval_set)

        train_ids = {example.example_id for example in train}
        eval_ids = {example.example_id for example in eval_set}
        train_features = {example.feature_hash for example in train}
        eval_features = {example.feature_hash for example in eval_set}

        id_overlap = len(train_ids & eval_ids)
        feature_overlap = len(train_features & eval_features)
        feature_overlap_ratio = feature_overlap / max(len(eval_features), 1)
        contaminated = (
            id_overlap > self.max_id_overlap
            or feature_overlap_ratio > self.max_feature_overlap_ratio
        )

        input_fingerprint = digest(
            {
                "train": self._dataset_fingerprint(train),
                "eval": self._dataset_fingerprint(eval_set),
            }
        )
        policy_fingerprint = digest(
            {
                "max_id_overlap": self.max_id_overlap,
                "max_feature_overlap_ratio": self.max_feature_overlap_ratio,
            }
        )
        report_body = {
            "input_fingerprint": input_fingerprint,
            "policy_fingerprint": policy_fingerprint,
            "id_overlap": id_overlap,
            "feature_overlap": feature_overlap,
            "feature_overlap_ratio": feature_overlap_ratio,
            "train_n": len(train),
            "eval_n": len(eval_set),
            "contaminated": contaminated,
        }
        return ContaminationReport(
            id_overlap=id_overlap,
            feature_overlap=feature_overlap,
            feature_overlap_ratio=feature_overlap_ratio,
            train_n=len(train),
            eval_n=len(eval_set),
            contaminated=contaminated,
            input_fingerprint=input_fingerprint,
            policy_fingerprint=policy_fingerprint,
            fingerprint=digest(report_body),
        )

File: go/poison/poison.go
package poison

import (
	"crypto/sha256"
	"encoding/hex"
	"encoding/json"
	"fmt"
	"math"
	"sort"
)

type Example struct {
	ID, FeatureHash string
}

type Report struct {
	IDOverlap          int
	FeatureOverlap     int
	FeatureOverlapRatio float64
	TrainN             int
	EvalN              int
	Contaminated       bool
	InputFingerprint   string
	PolicyFingerprint  string
	Fingerprint        string
	RefuseReason       string
}

func digest(value any) string {
	raw, _ := json.Marshal(value)
	hash := sha256.Sum256(raw)
	return hex.EncodeToString(hash[:])
}

func validateDataset(name string, examples []Example) error {
	seen := map[string]struct{}{}
	for _, example := range examples {
		if example.ID == "" {
			return fmt.Errorf("%s example ID must be non-empty", name)
		}
		if example.FeatureHash == "" {
			return fmt.Errorf("%s feature hash must be non-empty", name)
		}
		if _, exists := seen[example.ID]; exists {
			return fmt.Errorf("duplicate %s example ID: %s", name, example.ID)
		}
		seen[example.ID] = struct{}{}
	}
	return nil
}

func datasetFingerprint(examples []Example) [][2]string {
	pairs := make([][2]string, len(examples))
	for i, example := range examples {
		pairs[i] = [2]string{example.ID, example.FeatureHash}
	}
	sort.Slice(pairs, func(i, j int) bool {
		if pairs[i][0] == pairs[j][0] {
			return pairs[i][1] < pairs[j][1]
		}
		return pairs[i][0] < pairs[j][0]
	})
	return pairs
}

// AnalyzeChecked detects exact example-ID and exact feature-hash overlap under
// an explicit bounded policy. It does not itself compute semantic similarity.
func AnalyzeChecked(train, eval []Example, maxID int, maxFeatRatio float64) (Report, error) {
	if maxID < 0 {
		return Report{}, fmt.Errorf("maxID must be non-negative")
	}
	if math.IsNaN(maxFeatRatio) || math.IsInf(maxFeatRatio, 0) || maxFeatRatio < 0 || maxFeatRatio > 1 {
		return Report{}, fmt.Errorf("maxFeatRatio must be finite and in [0,1]")
	}
	if err := validateDataset("train", train); err != nil {
		return Report{}, err
	}
	if err := validateDataset("eval", eval); err != nil {
		return Report{}, err
	}

	trainIDs, evalIDs := map[string]struct{}{}, map[string]struct{}{}
	trainFeatures, evalFeatures := map[string]struct{}{}, map[string]struct{}{}
	for _, example := range train {
		trainIDs[example.ID] = struct{}{}
		trainFeatures[example.FeatureHash] = struct{}{}
	}
	for _, example := range eval {
		evalIDs[example.ID] = struct{}{}
		evalFeatures[example.FeatureHash] = struct{}{}
	}

	idOverlap, featureOverlap := 0, 0
	for id := range evalIDs {
		if _, exists := trainIDs[id]; exists {
			idOverlap++
		}
	}
	for feature := range evalFeatures {
		if _, exists := trainFeatures[feature]; exists {
			featureOverlap++
		}
	}
	featureRatio := 0.0
	if len(evalFeatures) > 0 {
		featureRatio = float64(featureOverlap) / float64(len(evalFeatures))
	}
	contaminated := idOverlap > maxID || featureRatio > maxFeatRatio

	inputFingerprint := digest(struct {
		Train [][2]string `json:"train"`
		Eval  [][2]string `json:"eval"`
	}{datasetFingerprint(train), datasetFingerprint(eval)})
	policyFingerprint := digest(struct {
		MaxID        int     `json:"max_id_overlap"`
		MaxFeatRatio float64 `json:"max_feature_overlap_ratio"`
	}{maxID, maxFeatRatio})
	reportFingerprint := digest(struct {
		InputFingerprint  string  `json:"input_fingerprint"`
		PolicyFingerprint string  `json:"policy_fingerprint"`
		IDOverlap         int     `json:"id_overlap"`
		FeatureOverlap    int     `json:"feature_overlap"`
		FeatureRatio      float64 `json:"feature_overlap_ratio"`
		TrainN            int     `json:"train_n"`
		EvalN             int     `json:"eval_n"`
		Contaminated      bool    `json:"contaminated"`
	}{inputFingerprint, policyFingerprint, idOverlap, featureOverlap, featureRatio, len(train), len(eval), contaminated})

	return Report{
		IDOverlap: idOverlap,
		FeatureOverlap: featureOverlap,
		FeatureOverlapRatio: featureRatio,
		TrainN: len(train),
		EvalN: len(eval),
		Contaminated: contaminated,
		InputFingerprint: inputFingerprint,
		PolicyFingerprint: policyFingerprint,
		Fingerprint: reportFingerprint,
	}, nil
}

// Analyze preserves the original API. Invalid requests fail closed as a
// contaminated report; callers needing the refusal reason should use AnalyzeChecked.
func Analyze(train, eval []Example, maxID int, maxFeatRatio float64) Report {
	report, err := AnalyzeChecked(train, eval, maxID, maxFeatRatio)
	if err != nil {
		return Report{Contaminated: true, RefuseReason: err.Error()}
	}
	return report
}
